api/org/commcare/api/json/JsonActionUtils.java

```diff
@@ -0,0 +1,270 @@
+     */
+    private static List<Pair<Integer, Integer>> stepToList(String index) {
+        ArrayList<Pair<Integer, Integer>> ret = new ArrayList<>();
+        String[] split = index.split(",");
+        List<String> list = Arrays.asList(split);
+        Collections.reverse(list);
+        for (String step : list) {
+            if (!step.trim().equals("")) {
+                Pair<Integer, Integer> pair = stepFromString(step);
+                ret.add(pair);
+            }
+        }
+        return ret;
+    }
+
+    /**
+     * Given the string representation of one "Step" in a form index, return an Integer pair of <index, multiplicity>
+     */
+    private static Pair<Integer, Integer> stepFromString(String step) {
+        // honestly not sure. thanks obama/drew
+        if (step.endsWith("J")) {
+            return new Pair<>(Integer.parseInt("" + step.substring(0, step.length() - 1)), TreeReference.INDEX_REPEAT_JUNCTURE);
+        }
+        // we want to deal with both '.' and '_' as separators for the time being for TF legacy reasons
+        String[] split = step.split("[._:]");
+        // the form index is the first part, the multiplicity is the second
+        int i = Integer.parseInt(split[0].trim());
+        int mult = -1;
+        if (split.length > 1 && split[1] != null) {
+            mult = Integer.parseInt(split[1].trim());
+        }
+        return new Pair<>(i, mult);
+    }
+
+    /**
+     * Given a list of steps (see above) to be traversed and a current Form index,
+     * pop the top step and create a new FormIndex from this step with the current as its parent, then recursively
+     * call this function with the remaining steps and the new FormIndex
+     */
+    private static FormIndex reduceFormIndex(List<Pair<Integer, Integer>> steps, FormIndex current) {
+        if (steps.size() == 0) {
+            return current;
+        }
+        Pair<Integer, Integer> currentStep = steps.remove(0);
+        FormIndex nextLevel = new FormIndex(current, currentStep.first, currentStep.second, null);
+        return reduceFormIndex(steps, nextLevel);
+    }
+
+    private static String[] parseMultiSelectString(String answer) {
+        answer = answer.trim();
+        if (answer.startsWith("[") && answer.endsWith("]")) {
+            answer = answer.substring(1, answer.length() - 1);
+        }
+        String[] ret = answer.split(" ");
+        for (int i = 0; i < ret.length; i++) {
+            ret[i] = ret[i].replace(",", "");
+        }
+        return ret;
+    }
+
+    public static JSONArray walkToJSON(FormEntryModel fem, FormEntryController fec) {
+        try {
+            JSONArray ret = new JSONArray();
+            FormIndex formIndex = FormIndex.createBeginningOfFormIndex();
+            Walker walker = new Walker(ret, formIndex, fec, fem);
+            walker.walk();
+            return ret;
+        } catch (Exception e) {
+            e.printStackTrace();
+            return null;
```

Comment:
There are tons of things that could have gone wrong in this call, and returning null from here will both crash or break other code, and has no clear semantic meaning. You need to figure out what can be failing here and make sure that it is handled